doc/development/pluginsystem.rst

```diff
@@ -0,0 +1,210 @@
+FoamAdapter's plugin architecture is designed to enable users and developers to add new physics models, boundary conditions, and solver modules without modifying the core codebase.
+This approach promotes maintainability, collaboration, and rapid prototyping of new features.
+
+Overview
+--------
+
+The PluginSystem provides a runtime-extensible configuration system using Pydantic discriminated unions and a registry pattern.
+The system allows to register child classes on numerous base classes, each with its own registry.
+
+Core Components
+---------------
+
+PluginRegistry Dataclass
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``PluginRegistry`` dataclass stores metadata for each plugin base type:
+
+* ``base_cls``: The plugin base class (Pydantic BaseModel)
+* ``discriminator_variable``: Field name holding the union (e.g., 'shape', 'plugin')
+* ``discriminator``: Discriminator field name in plugin configs (e.g., 'shape_type', 'plugin_type')
+* ``plugin_registry``: List of registered plugin configuration classes
+* ``plugin_model``: Dynamically generated extensible Pydantic model
+
+PluginSystem Class
+~~~~~~~~~~~~~~~~~~
+
+The central registry stores all plugin families within the class dictionary variable ``_registry`` .
+Each key represents a plugin family name mapping to its corresponding PluginRegistry instance.
+
+Implementation Details
+----------------------
+
+Registration Process
+~~~~~~~~~~~~~~~~~~~~
+
+The decorator-based registration follows a two-step process:
+
+1. **Base Class Registration**: ``@PluginSystem.register()`` creates a registry entry and adds helper methods
+2. **Plugin Registration**: ``@BaseClass.register`` adds plugin classes to the registry and regenerates the union model
+
+Dynamic Model Generation
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The system uses ``pydantic.create_model()`` to dynamically generate extensible models.
+Each registration updates the discriminated union type and recreates the model.
+The discriminator field enables automatic deserialization based on the type identifier.
+
+Usage Examples
+--------------
+
+Basic Plugin Setup
+~~~~~~~~~~~~~~~~~~
+
+The following example demonstrates how to define a plugin base class and register multiple plugin configurations.
+
+.. code-block:: python
+
+    from pydantic import BaseModel
+    from typing import Literal
+    from foamadapter.core.plugin_system import PluginSystem
+
+    @PluginSystem.register(discriminator_variable="shape", discriminator="shape_type")
+    class ShapeInterface(BaseModel):
+        color: str
+        name: str = "default"
+
+    @ShapeInterface.register
+    class CircleConfig(BaseModel):
+        shape_type: Literal["circle"]
+        radius: float
+
+    @ShapeInterface.register
+    class SquareConfig(BaseModel):
+        shape_type: Literal["square"]
+        side: float
+
+Creating and Using Configurations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A usage example creating and using plugin configurations is shown below.
```

Comment:
The example below only shows how to use/instantiate the plugin. Also i think it needs a bit of elaboration what the "plugin" in this example is. I.e. how to instantiate plugins of the Shape base class?